Modded Emote Bubble

Cyrillya opened the pull request:
### What is the new feature?
Added stuff for modded emotes.

### Why should this be part of tModLoader?
There was a issue opened about 3 years ago: #689
This pr makes that suggestion come true.
As 1.4 implemented player-usable emotes, the emote system is built and mod emotes should be a thing now.

### Sample usage for the new feature
Included in the Example Mod examples added.
There is also a feature showcase here: https://imgur.com/iutYGTl

### ExampleMod updates
- Emote bubbles for
  - Example mod NPCs
  - Example biome
  - Minion boss
  - Example pickaxe (pops out when players use example pickaxe)
  - Example item (pops out when ExZombieThief steals a ExItem)
- Example person reacts with angry emotes more often against the demolitionist, and minion boss emote against other NPCs

Cyrillya committed:
Basic emote bubble things + example

Cyrillya committed:
ModEmoteBubble stuff completed

direwolf420 commented on ExampleMod/Common/GlobalNPCs/EmotePickerGlobalNPC.cs:
```diff
@@ -0,0 +1,28 @@
+{
+	// This is where we add our regular emotes for all NPCs
+	public class EmotePickerGlobalNPC : GlobalNPC
+	{
+		public override int PickEmote(NPC npc, Player closetPlayer, List<int> emoteList, WorldUIAnchor otherAnchor) {
+			// Add the biome emote to the list if the player is in Example Biome
```
closetPlayer -> closestPlayer

direwolf420 commented on ExampleMod/Common/GlobalNPCs/EmotePickerGlobalNPC.cs:
```diff
@@ -0,0 +1,28 @@
+	{
+		public override int PickEmote(NPC npc, Player closetPlayer, List<int> emoteList, WorldUIAnchor otherAnchor) {
+			// Add the biome emote to the list if the player is in Example Biome
+			// And with random chance, this emote will be less likely to appear
+			if (Main.rand.NextBool(2) && ModContent.GetInstance<ExampleSurfaceBiome>().IsBiomeActive(closetPlayer))
+				emoteList.Add(ModContent.EmoteBubbleType<ExampleBiomeEmote>());
+
```
This and every other single-line block statement (**in ExampleMod only!**): Change to 
```
condition {
    code
}
```

I won't be highlighting the other ones, there's too many and it would get quite bloated

direwolf420 commented on ExampleMod/Content/EmoteBubbles/ExampleBiomeEmote.cs:
```diff
@@ -0,0 +1,53 @@
+	// It performs totally the same as vanilla.
+	// Check Common/GlobalNPC/EmotePickerGlobalNPC.cs for adding this emote for all NPCs.
+	public class ExampleBiomeEmote : ModEmoteBubble
+	{
+		public override void SetStaticDefaults() => EmoteName.SetDefault("exbiome");
+
```
single line rule applies here too

direwolf420 commented on ExampleMod/Content/Items/Tools/ExamplePickaxe.cs:
```diff
@@ -38,6 +40,16 @@ public class ExamplePickaxe : ModItem
+				// Note that we don't have to sync the emote because UseAnimation runs on all clients and server
+				EmoteBubble.MakePlayerEmote(player, ModContent.EmoteBubbleType<ExamplePickaxeEmote>(), false);
+			}
+
+			base.UseAnimation(player);
+		}
```
base calls in ExampleMod is usually not included if it doesn't do anything

direwolf420 commented on patches/tModLoader/Terraria/ModLoader/GlobalNPC.cs:
```diff
@@ -706,5 +707,16 @@ public virtual void DrawBehind(NPC npc, int index)
+		/// <param name="npc"></param>
+		/// <param name="emoteList">A list of emote IDs from which the NPC will randomly select one</param>
+		/// <param name="otherAnchor">A <see cref="WorldUIAnchor"/> instance that indicates the target of this emote conversation. Use this to get the instance of the <see cref="NPC"/> or <see cref="Player"/> this NPC is talking to.</param>
+		/// <returns>Return -1 to use vanilla mechanic (pick one from the list), otherwise pick the emote by the returned ID. If the ID does not correspond to an emote then it will prevent the emote from being used. Returns -1 by default</returns>
+		public virtual int PickEmote(NPC npc, Player closetPlayer, List<int> emoteList, WorldUIAnchor otherAnchor) {
+			return -1;
```
closetPlayer -> closestPlayer, also missing documentation

direwolf420 reviewed (commented):
Thanks for the PR!
Here's some very basic formatting tweaks

Question: Is it possible to add your emotes to vanilla categories (i.e. I add a new critter + associated emote and want to have it selectable under the "Critters & Monsters" category)? And maybe also keep it under the modded category, not just move it.

Cyrillya committed:
- Format fixes
- globalEmotes don't clear during Unload fixed
- You can add mod emotes to vanilla categories now

Cyrillya commented:
Formatting issues are fixed.
Mod emotes can now be added to vanilla categories via the  `AddToCategory()` method. I was considering `EmoteID.Sets`, but that won't work for vanilla boss emotes (they only unlock when the corresponding boss is downed).

Cyrillya commented:
Formatting issues are fixed.
> Question: Is it possible to add your emotes to vanilla categories (i.e. I add a new critter + associated emote and want to have it selectable under the "Critters & Monsters" category)? And maybe also keep it under the modded category, not just move it.

Mod emotes can now be added to vanilla categories via the `AddToCategory()` method. I was considering `EmoteID.Sets`, but that won't work for vanilla boss emotes (they only unlock when the corresponding boss is downed).

Cyrillya committed:
Make `UIEmotesMenu.GetEmotesXX()` `public static` in case someone needs them

Cyrillya committed:
Gamepad support

Cyrillya committed:
More comments

Cyrillya committed:
ExItemEmote that shows when ExZombieThief steals a ExItem

Cyrillya commented:
Example for enemies using emotes
![ZombieThief](https://user-images.githubusercontent.com/35227653/190849620-f3b2fd12-0f2e-4f6f-8eb5-99a77f511569.gif)

direwolf420 commented on patches/tModLoader/Terraria/GameContent/UI/States/UIEmotesMenu.cs.patch:
```diff
@@ -0,0 +1,178 @@
+ 				148
+ 			};
++
++			if (EmoteBubbleLoader.categoryEmoteLookup.TryGetValue(EmoteID.Category.General, out var modEmotes)) {
++				list.AddRange(from e in modEmotes where e.IsUnlocked() select e.Type);
++			}
```
This and the other similar code in the GetEmotesX methods should get condensed into a single call to a new (internal) helper method in EmoteBubbleLoader.

Cyrillya replied:
Fixed

Cyrillya committed:
Add `AddEmotesToCategory` method

direwolf420 commented:
Making it an extension is not the best idea I think

Cyrillya commented:
> Making it an extension is not the best idea I think

Should I use a static method and call it multiple times for all categories?

direwolf420 commented:
Ok I checked again and since it keeps patches minimal, it should be fine as is

Cyrillya committed:
Extension method comment

Cyrillya committed:
Resolve conflicts

Cyrillya committed:
Merge branch '1.4' into 1.4-mod-emote

Chicken-Bones commented on ExampleMod/Content/NPCs/ExamplePerson.cs:
```diff
@@ -356,6 +363,24 @@ public class ExamplePerson : ModNPC
+
+		// Let the NPC "talk about" minion boss
+		public override int PickEmote(Player closestPlayer, List<int> emoteList, WorldUIAnchor otherAnchor) {
+			// By default this NPC will always use the Minion Boss Emote even if Minion Boss is not downed yet
```
"always" is not necessarily correct, because it depends what's in the list.

Chicken-Bones commented on patches/tModLoader/Terraria/Chat/Commands/EmojiCommand.cs.patch:
```diff
@@ -0,0 +1,38 @@
+ 		public void Initialize() {
+ 			_byName.Clear();
+-			for (int i = 0; i < 151; i++) {
++			for (int i = 0; i < EmoteID.Count; i++) {
```
Reason for change?

Chicken-Bones commented on patches/tModLoader/Terraria/Initializers/ChatInitializer.cs.patch:
```diff
@@ -21,3 +29,21 @@
+ 
+ 		public static void PrepareAliases() {
+ 			ChatManager.Commands.ClearAliases();
+-			for (int i = 0; i < 151; i++) {
++			for (int i = 0; i < EmoteID.Count; i++) {
```
Lets not change if we don't need to

Chicken-Bones commented on patches/tModLoader/Terraria/GameContent/UI/States/UIEmotesMenu.cs.patch:
```diff
@@ -0,0 +1,144 @@
+ 
+ 			if (DD2Event.DownedInvasionAnyDifficulty)
+ 				list.Add(133);
+-
++				
```
blank line change

Chicken-Bones commented on ExampleMod/Content/EmoteBubbles/MinionBossEmote.cs:
```diff
@@ -0,0 +1,24 @@
+	public class MinionBossEmote : ModEmoteBubble
+	{
+		public override void SetStaticDefaults() {
+			// This is the emote command name for this emote.
+			// You will want to lowercase all letters to match the vanilla command.
+			// If you don't have this set, it will be the lowercase name of the class.
+			EmoteName.SetDefault("minionboss");
```
@JavidPack I believe we're moving away from `SetDefault` in examples, in favour of just using the translation file.

JavidPack replied:
We are...but still haven't been too consistent with doing it. At the very least we sometimes leave a comment with the old way and direct users to the localization files. I would only keep the comment in 1 example, the example that has comments and is intended to be the teaching example.

Chicken-Bones commented on patches/tModLoader/Terraria/GameContent/UI/States/UIEmotesMenu.cs.patch:
```diff
@@ -0,0 +1,144 @@
+ 				_list.Add(new EmotesGroupListItem(title, currentGroupIndex++, maxEmotesPerRow, emoteIds.ToArray()));
+ 		}
+ 
+-		private List<int> GetEmotesGeneral() {
```
You could get better patches by editing the `TryAddingList` lines instead

Cyrillya replied:
Editing the `TryAddingList` line will cause modders to get only vanilla emotes when they want to get a list of emotes for a category.

Chicken-Bones commented on ExampleMod/Content/Items/Tools/ExamplePickaxe.cs:
```diff
@@ -38,6 +40,14 @@ public class ExamplePickaxe : ModItem
+		public override void UseAnimation(Player player) {
+			// Randomly causes the player to use Example Pickaxe Emote when using the item
+			if (Main.rand.NextBool(20)) {
+				// Note that we don't have to sync the emote because UseAnimation runs on all clients and server
+				EmoteBubble.MakePlayerEmote(player, ModContent.EmoteBubbleType<ExamplePickaxeEmote>(), false);
```
The value of `Main.rand` will not be the same on all clients, so you should only do this on one machine and sync the emote so all players see it at the same time

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/ModEmoteBubble.cs:
```diff
@@ -0,0 +1,153 @@
+		/// <summary>
+		/// Gets called when your emote bubble spawns in world.
+		/// </summary>
+		public virtual void OnSpawn() { }
```
Should this be `OnSpawn`? Thoughts @JavidPack

JavidPack replied:
Sounds fine to me, other things that enter the world use OnSpawn

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/ModEmoteBubble.cs:
```diff
@@ -0,0 +1,153 @@
+			ModContent.Request<Texture2D>(Texture);
+			SetStaticDefaults();
+
+			if (EmoteName.IsDefault())
+				EmoteName.SetDefault(Name.ToLower()); // It will be the lowercase name of the class by default.
```
Not sure if this is a good idea, we want to encourage use of localisation files.

JavidPack replied:
This is consistent with all other content, I think it is fine to assume english name from classname

Chicken-Bones reviewed (changes requested):
Incredible work! So many things are bang on.

I've reviewed everything except for `EmoteBubble.cs` and `EmoteBubbleLoader.cs` because I ran out of time, will get to those tomorrow.

Chicken-Bones commented on patches/tModLoader/Terraria/GameContent/UI/EmoteBubble.cs.patch:
```diff
@@ -1,19 +1,42 @@
+-		public static Dictionary<int, EmoteBubble> byID = new Dictionary<int, EmoteBubble>();
 +		internal static int[] CountNPCs = new int[NPCID.Count];
- 		public static Dictionary<int, EmoteBubble> byID = new Dictionary<int, EmoteBubble>();
++		internal static Dictionary<int, EmoteBubble> byID = new Dictionary<int, EmoteBubble>(); // TML: Changed from public to internal. Use EmoteBubbles instead for code readability
```
generally no need to comment on the reason for changing stuff `public` or `internal`

Chicken-Bones replied:
This specific one should stay `private` though

Cyrillya replied:
> This specific one should stay `private` though

It's used in `MessageBuffer.cs` as well, I will keep it `internal`

Chicken-Bones commented on patches/tModLoader/Terraria/GameContent/UI/EmoteBubble.TML.cs:
```diff
@@ -0,0 +1,61 @@
+		/// <summary>
+		/// A dictionary contains all the existing emote bubbles in the world
+		/// </summary>
+		public static ref Dictionary<int, EmoteBubble> EmoteBubbles => ref byID;
```
I don't think this should be exposed, you've given people all the helpers they need. If you think there's a good reason, to expose it, you should do it as:
```
public static IReadOnlyDictionary<int, EmoteBubble> EmoteBubbles
```

Chicken-Bones commented on patches/tModLoader/Terraria/GameContent/UI/EmoteBubble.TML.cs:
```diff
@@ -0,0 +1,61 @@
+		/// </summary>
+		/// <param name="whoAmI"></param>
+		/// <returns></returns>
+		public static EmoteBubble GetExistingEmoteBubble(int whoAmI) => EmoteBubbles.GetValueOrDefault(whoAmI, null);
```
No need for `, null` here

Chicken-Bones commented on patches/tModLoader/Terraria/GameContent/UI/EmoteBubble.TML.cs:
```diff
@@ -0,0 +1,61 @@
+		/// Whether or not this emote is displaying
+		/// <br>The first and the last 6 frames are for bubble-popping animation. The emote content is only displayed after the animation</br>
+		/// </summary>
+		public bool IsDisplayingEmote => lifeTime < 6 || lifeTimeStart - lifeTime < 6;
```
Suggest inverting, and calling it `IsFullyDisplayed`

Chicken-Bones commented on patches/tModLoader/Terraria/GameContent/UI/EmoteBubble.TML.cs:
```diff
@@ -0,0 +1,61 @@
+		/// <param name="type"></param>
+		/// <returns>Returns the <see cref="WhoAmI"/> of the <see cref="EmoteBubble"/>. Returns -1 if there is none</returns>
+		public static int TryGetFirst(int type) {
+			var emote = EmoteBubbles.First(i => i.Value.emote == type);
```
You would need to use `FirstOrDefault` here.

Chicken-Bones commented on patches/tModLoader/Terraria/GameContent/UI/EmoteBubble.TML.cs:
```diff
@@ -0,0 +1,61 @@
+		/// </summary>
+		/// <param name="type"></param>
+		/// <returns>Returns the <see cref="WhoAmI"/> of the <see cref="EmoteBubble"/>. Returns -1 if there is none</returns>
+		public static int TryGetFirst(int type) {
```
What's the use case for this one?

Chicken-Bones commented on patches/tModLoader/Terraria/GameContent/UI/EmoteBubble.cs.patch:
```diff
@@ -1,19 +1,42 @@
+-		public static int NextID;
+-		public int ID;
++		private static int NextID; // TML: Changed from public to private
++		internal int ID; // TML: Changed from public to internal
```
No need for this access change

Cyrillya replied:
This is for notifying modders to use `emote` field for real emote ID and avoid confusing. (The `ID` field here is like `whoAmI` in NPC)

Chicken-Bones commented on patches/tModLoader/Terraria/GameContent/UI/EmoteBubble.cs.patch:
```diff
@@ -71,6 +94,189 @@
++			if (!EmoteBubbleLoader.UpdateFrame(this)) {
++				return;
++			}
++			// TML: Fixed frameSpeed not being used, might be a bug?
```
Not a bug, the loss of `const` field references is a result of compilation+decompilation.

Chicken-Bones commented on patches/tModLoader/Terraria/GameContent/UI/EmoteBubble.cs.patch:
```diff
@@ -71,6 +94,189 @@
++			}
+ 			emoteBubble.ID = AssignNewID();
+ 			byID[emoteBubble.ID] = emoteBubble;
+-			emoteBubble.PickNPCEmote(other);
```
comment out instead of removing please

Chicken-Bones commented on patches/tModLoader/Terraria/GameContent/UI/EmoteBubble.cs.patch:
```diff
@@ -71,6 +94,189 @@
+ 		}
+ 
+ 		public static int AssignNewID() => NextID++;
+-
++		
```
accidental whitespace change

Chicken-Bones commented on patches/tModLoader/Terraria/GameContent/UI/EmoteBubble.cs.patch:
```diff
@@ -71,6 +94,189 @@
+ 			SpriteEffects effect = SpriteEffects.None;
+ 			Vector2 vector = GetPosition(out effect);
+ 			vector = vector.Floor();
+-			bool flag = lifeTime < 6 || lifeTimeStart - lifeTime < 6;
+-			Rectangle value2 = value.Frame(8, 39, (!flag) ? 1 : 0);
++			bool flag = IsDisplayingEmote;
++			Rectangle value2 = bubbleTexture.Frame(8, 39, (!flag) ? 1 : 0);
```
No need for these patches

Chicken-Bones commented on patches/tModLoader/Terraria/GameContent/UI/EmoteBubble.cs.patch:
```diff
@@ -71,6 +94,189 @@
+ 
+ 		private void Draw(SpriteBatch sb) {
+ 			Texture2D value = TextureAssets.Extra[48].Value;
++			Texture2D bubbleTexture = TextureAssets.Extra[48].Value;
```
Suggest `bubbleTexture = value`, could be moved down below, to where `ModEmoteBubble` kicks in

Chicken-Bones commented on patches/tModLoader/Terraria/GameContent/UI/EmoteBubble.cs.patch:
```diff
@@ -71,6 +94,189 @@
+ 			}
+ 
++			other ??= new WorldUIAnchor();
++			int modPickedEmote = NPCLoader.PickEmote((NPC)anchor.entity, plr, list, other);
```
This should change to return `int?`
- `null` -> vanilla behavior
- `-1` -> no emote
- `int i` -> emote id `i`

Chicken-Bones replied:
No need for the `out` variable then

Chicken-Bones commented on patches/tModLoader/Terraria/GameContent/UI/EmoteBubble.cs.patch:
```diff
@@ -71,6 +94,189 @@
+ 		public static int NewBubbleNPC(WorldUIAnchor bubbleAnchor, int time, WorldUIAnchor other = null) {
+ 			EmoteBubble emoteBubble = new EmoteBubble(0, bubbleAnchor, time);
++			emoteBubble.PickNPCEmote(out bool useEmote, other);
++			if (!useEmote) {
```
With suggested change to `NewBubbleNPC` below, can be:
```
if (emote < 0)
    return -1;
```

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/EmoteBubbleLoader.cs:
```diff
@@ -0,0 +1,174 @@
+
+		internal static Dictionary<Mod, List<int>> GetAllUnlockedModEmotes() {
+			var result = new Dictionary<Mod, List<int>>();
+			foreach (var modEmoteBubble in from modEmote in emoteBubbles where modEmote.IsUnlocked() select modEmote)
```
Please use 'functional style' LINQ in tML. 
```cs
foreach (var modEmote in emoteBubbles.Where(e => e.IsUnlocked())
```

Alternatively, since it's just a simple filter, put it in the loop:
```cs
foreach (var modEmote in emoteBubbles) {
    if (!modeEmote.isUnlocked()) continue;
    ...
```

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/EmoteBubbleLoader.cs:
```diff
@@ -0,0 +1,174 @@
+			var result = new Dictionary<Mod, List<int>>();
+			foreach (var modEmoteBubble in from modEmote in emoteBubbles where modEmote.IsUnlocked() select modEmote)
+			{
+				if (!result.TryGetValue(modEmoteBubble.Mod, out var emoteList)) {
+					result[modEmoteBubble.Mod] = new List<int> { modEmoteBubble.Type };
+					continue;
```
Preferred idiom is:
```cs
				if (!result.TryGetValue(modEmoteBubble.Mod, out var emoteList))
					result[modEmoteBubble.Mod] = emoteList =  new();
```

Then the `emoteList.Add(...)` call works either way

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/ModEmoteBubble.cs:
```diff
@@ -0,0 +1,153 @@
+		/// <summary>
+		/// This is the <see cref="EmoteBubble"/> instance.
+		/// </summary>
+		public EmoteBubble EmoteBubble { get; internal set; }
```
This should just be an accessor for the `Entity` property, see `ModItem` for an example

`public Item EmoteBubble => Entity;`

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/EmoteBubbleLoader.cs:
```diff
@@ -0,0 +1,174 @@
+			if (result && GetEmoteBubble(emoteType)?.UpdateFrameInEmoteMenu(ref frameCounter) is false) {
+				result = false;
+			}
+			return result;
```
Perhaps this, and all the ones below would be better written as:
```cs
if (!result) return false;

return GetEmoteBubble(emoteType)?.UpdateFrameInEmoteMenu(ref frameCounter) ?? true;
```

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/EmoteBubbleLoader.cs:
```diff
@@ -0,0 +1,174 @@
+		}
+
+
+		public static Rectangle? GetFrame(EmoteBubble emoteBubble) {
```
Should this method just employ early returns?

Chicken-Bones reviewed (changes requested):
Alright, got some more in depth feedback for you. Please come to the Discord if you'd like to discuss anything

Cyrillya committed:
Merge branch 'tModLoader:1.4' into 1.4-mod-emote

Cyrillya committed:
Code optimization in tModLoader side

Cyrillya committed:
ExEmotes name use .hjson files + Fix emote command name translation not applied

Cyrillya committed:
Net sync fix for pickaxe emote

Cyrillya committed:
Delete `TryGetFirst` because it is barely used (unnecessary)

Cyrillya committed:
Use plr.ZoneForest

Cyrillya committed:
Fix conflict

JavidPack committed:
Merge branch '1.4.4' into pr/2918

JavidPack committed:
Fix patches, adjust to localization and other changes

JavidPack committed:
File-scoped namespaces

JavidPack committed:
Patch cleanup

JavidPack committed:
More patch fixes

JavidPack commented:
Ported to 1.4.4, things seem to be working correctly, but I'm not too familiar with everything that is possible with emotes, so I'm not sure.

direwolf420 commented on ExampleMod/Content/EmoteBubbles/MinionBossEmote.cs:
```diff
@@ -0,0 +1,25 @@
+	public class MinionBossEmote : ModEmoteBubble
+	{
+		public override void SetStaticDefaults() {
+			// As for emote command name, you will want to lowercase all letters to match the vanilla command.
+			// If you don't have this set, it will be the lowercase name of the class.
+			// EmoteName automatically assigned from localization files (files in Localization/),
+			// but the commented line below is the normal approach.
+			// EmoteName.SetDefault("exminionboss");
```
This needs updating for new 1.4.4 localization

JavidPack committed:
Fix some comments

Cyrillya commented:
I think it is fine and ready to be merged

JavidPack committed:
Rename CommandName to Command, fix MP sync, fix issue switching languages/lowercase
